community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputIterator.java

```diff
@@ -0,0 +1,135 @@
+import org.neo4j.csv.reader.Readables;
+import org.neo4j.csv.reader.Source;
+import org.neo4j.csv.reader.Source.Chunk;
+import org.neo4j.csv.reader.SourceTraceability;
+import org.neo4j.unsafe.impl.batchimport.input.Collector;
+import org.neo4j.unsafe.impl.batchimport.input.Groups;
+import org.neo4j.unsafe.impl.batchimport.input.InputChunk;
+
+import static java.util.Arrays.copyOf;
+
+/**
+ * Iterates over one stream of data, where all data items conform to the same {@link Header}.
+ * Typically created from {@link CsvGroupInputIterator}.
+ */
+class CsvInputIterator implements SourceTraceability, Closeable
+{
+    private final CharReadable stream;
+    private final Chunker chunker;
+    private final Header header;
+    private final Decorator decorator;
+    private final Configuration config;
+
+    CsvInputIterator( CharReadable stream, Decorator decorator, Header header, Configuration config, IdType idType, Collector badCollector,
+            Extractors extractors )
+    {
+        this.stream = stream;
+        this.config = config;
+        this.decorator = decorator;
+        this.header = header;
+        this.chunker = config.multilineFields()
+                ? new EagerParserChunker( stream, idType, header, badCollector, extractors, 1_000, config, decorator )
+                : new ClosestNewLineChunker( stream, config.bufferSize() );
+    }
+
+    CsvInputIterator( CharReadable stream, Decorator decorator, Header.Factory headerFactory, IdType idType,
+            Configuration config, Groups groups, Collector badCollector, Extractors extractors ) throws IOException
+    {
+        this( stream, decorator, extractHeader( stream, headerFactory, idType, config, groups ), config, idType, badCollector, extractors );
+    }
+
+    static Header extractHeader( CharReadable stream, Header.Factory headerFactory, IdType idType,
+            Configuration config, Groups groups ) throws IOException
+    {
+        if ( !headerFactory.isDefined() )
+        {
+            char[] firstLineBuffer = Readables.extractFirstLineFrom( stream );
+            // make the chunk slightly bigger than the header to not have the seeker think that it's reading
+            // a value bigger than its max buffer size
+            ChunkImpl firstChunk = new ChunkImpl( copyOf( firstLineBuffer, firstLineBuffer.length + 1 ) );
+            firstChunk.initialize( firstLineBuffer.length, stream.sourceDescription() );
+            CharSeeker firstSeeker = seeker( firstChunk, config );
+            return headerFactory.create( firstSeeker, config, idType, groups );
+        }
+
+        return headerFactory.create( null, null, null, null );
+    }
+
+    public boolean next( InputChunk chunk ) throws IOException
+    {
+        if ( config.multilineFields() )
+        {
+            EagerlyReadInputChunk csvChunk = (EagerlyReadInputChunk) chunk;
+            return chunker.nextChunk( csvChunk );
+        }
+
+        CsvInputChunk csvChunk = (CsvInputChunk) chunk;
```

Comment:
This will be slightly rewritten, but as with an API which has explicit allocation like:

````
MyInterface
{
    MyThing allocateAThing();

    void fillMyThinkWithSomething( MyThing thing );
}
````

This cast comes by design, doesn't it?